src/js/components/settings/roles.js

```diff
@@ -91,70 +121,133 @@ export class RoleManagement extends React.Component {
+                  target="_blank"
+                  disabled={!(name && (allowUserManagement || groups.some(group => group.selected)))}
+                  onClick={() => self.onSubmit()}
+                >
+                  Submit
+                </Button>
+              </div>
+            </>
+          )}
+          {editing && (
+            <>
+              <h4 style={{ marginTop: 5 }}>Edit the role</h4>
+              <TextField
+                label="Role name"
+                id="edit-role-name"
+                value={name}
+                disabled={true}
+                onChange={e => self.setState({ name: e.target.value })}
+                style={{ marginTop: 0, marginRight: 30 }}
+              />
+              <TextField
+                label="Description"
+                id="edit-role-description"
+                value={description}
+                placeholder="-"
+                onChange={e => self.setState({ description: e.target.value })}
+                style={{ marginTop: 0, marginRight: 30 }}
+              />
+              <div>
                 <FormControlLabel
-                  style={{ marginTop: 0, marginLeft: 0 }}
-                  key={group.name}
-                  control={<Checkbox color="primary" checked={group.selected} onChange={(e, checked) => self.handleGroupSelection(checked, group)} />}
-                  label={group.name}
+                  control={<Checkbox color="primary" onChange={(e, checked) => self.setState({ allowUserManagement: checked })} />}
+                  checked={allowUserManagement}
+                  label="Allow to manage other users"
                 />
-              ))}
-            </div>
+              </div>
+              {!!groups.length && (
+                <div className="flexbox column margin-top-small">
+                  <div>Device group permission</div>
+                  {groups.map(group => (
+                    <FormControlLabel
+                      style={{ marginTop: 0, marginLeft: 0 }}
+                      key={group.name}
+                      control={<Checkbox color="primary" checked={group.selected} onChange={(e, checked) => self.handleGroupSelection(checked, group)} />}
+                      label={group.name}
+                    />
+                  ))}
+                </div>
+              )}
+              <div className="flexbox centered" style={{ justifyContent: 'flex-end' }}>
+                <Button onClick={() => self.onCancel()} style={{ marginRight: 15 }}>
+                  Cancel
+                </Button>
+                <Button
+                  color="secondary"
+                  variant="contained"
+                  target="_blank"
+                  disabled={!(name && (allowUserManagement || groups.some(group => group.selected)))}
+                  onClick={() => self.onSubmit()}
+                >
+                  Submit
+                </Button>
+              </div>
+            </>
           )}
```

Comment:
@tranchitella due to the similarity in both cases I was more aiming at something like this:

```javascript
<Collapse in={adding || editing} className="margin-right-small filter-wrapper" classes={{ wrapperInner: 'margin-bottom-small margin-right' }}>
  <>
    <h4 style={{ marginTop: 5 }}>{adding ? 'Add a' : 'Edit the'} role</h4>
    <TextField
      label="Role name"
      id="role-name"
      value={name}
      disabled={editing}
      onChange={e => self.setState({ name: e.target.value })}
      style={{ marginTop: 0, marginRight: 30 }}
    />
    <TextField
      label="Description"
      id="role-description"
      value={description}
      placeholder="-"
      onChange={e => self.setState({ description: e.target.value })}
      style={{ marginTop: 0, marginRight: 30 }}
    />
    <div>
      <FormControlLabel
        control={<Checkbox color="primary" onChange={(e, checked) => self.setState({ allowUserManagement: checked })} />}
        checked={allowUserManagement}
        label="Allow to manage other users"
      />
    </div>
    {!!groups.length && (
      <div className="flexbox column margin-top-small">
        <div>Device group permission</div>
        {groups.map(group => (
          <FormControlLabel
            style={{ marginTop: 0, marginLeft: 0 }}
            key={group.name}
            control={<Checkbox color="primary" checked={group.selected} onChange={(e, checked) => self.handleGroupSelection(checked, group)} />}
            label={group.name}
          />
        ))}
      </div>
    )}
    <div className="flexbox centered" style={{ justifyContent: 'flex-end' }}>
      <Button onClick={() => self.onCancel()} style={{ marginRight: 15 }}>
        Cancel
      </Button>
      <Button
        color="secondary"
        variant="contained"
        target="_blank"
        disabled={!(name && (allowUserManagement || groups.some(group => group.selected)))}
        onClick={() => self.onSubmit()}
      >
        Submit
      </Button>
    </div>
  </>
</Collapse>
```

which should increase the shared parts - or you could also extract the adding/ editing sections into a parameterized component (which could result in a bit more work though)

Reply:
Yup, you are right, I've updated my PR